[Bug] Intermittent debug(T) test failure: org.mozilla.fenix.share.ShareControllerTest handleShareToApp should start a new sharing activity and close this

Log: https://treeherder.mozilla.org/logviewer.html#?job_id=319444807&repo=fenix

```
[task 2020-10-22T21:33:50.397Z] SUITE: org.mozilla.fenix.share.ShareControllerTest
[task 2020-10-22T21:33:50.397Z]   TEST: getShareSubject will return "shareSubject" if that is non null
[task 2020-10-22T21:33:50.397Z]     [Robolectric] org.mozilla.fenix.share.ShareControllerTest.getShareSubject will return "shareSubject" if that is non null: sdk=28; resources=BINARY
[task 2020-10-22T21:33:50.497Z]     I/MonitoringInstr: Instrumentation started!
[task 2020-10-22T21:33:51.397Z]   SUCCESS
[task 2020-10-22T21:33:51.397Z] 
[task 2020-10-22T21:33:51.397Z]   TEST: handleShareToApp should start a new sharing activity and close this
[task 2020-10-22T21:33:51.397Z]     [Robolectric] org.mozilla.fenix.share.ShareControllerTest.handleShareToApp should start a new sharing activity and close this: sdk=28; resources=BINARY
[task 2020-10-22T21:33:51.397Z]     I/MonitoringInstr: Instrumentation started!
[task 2020-10-22T21:33:51.497Z]   FAILURE
[task 2020-10-22T21:33:51.497Z] 
[task 2020-10-22T21:33:51.497Z] java.lang.AssertionError: Verification failed: call 1 of 1: RecentAppsStorage(#509).updateRecentApp(eq(activity))) was not called
[task 2020-10-22T21:33:51.497Z] 	at io.mockk.impl.recording.states.VerifyingState.failIfNotPassed(VerifyingState.kt:66)
[task 2020-10-22T21:33:51.497Z] 	at io.mockk.impl.recording.states.VerifyingState.recordingDone(VerifyingState.kt:42)
[task 2020-10-22T21:33:51.497Z] 	at io.mockk.impl.recording.CommonCallRecorder.done(CommonCallRecorder.kt:47)
[task 2020-10-22T21:33:51.497Z] 	at io.mockk.impl.eval.RecordedBlockEvaluator.record(RecordedBlockEvaluator.kt:60)
[task 2020-10-22T21:33:51.497Z] 	at io.mockk.impl.eval.VerifyBlockEvaluator.verify(VerifyBlockEvaluator.kt:30)
[task 2020-10-22T21:33:51.497Z] 	at io.mockk.MockKDsl.internalVerify(API.kt:118)
[task 2020-10-22T21:33:51.498Z] 	at io.mockk.MockKKt.verify(MockK.kt:146)
[task 2020-10-22T21:33:51.498Z] 	at io.mockk.MockKKt.verify$default(MockK.kt:143)
[task 2020-10-22T21:33:51.498Z] 	at org.mozilla.fenix.share.ShareControllerTest$handleShareToApp should start a new sharing activity and close this$1.invokeSuspend(ShareControllerTest.kt:113)
[task 2020-10-22T21:33:51.498Z] 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
[task 2020-10-22T21:33:51.498Z] 	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
[task 2020-10-22T21:33:51.498Z] 	at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:274)
[task 2020-10-22T21:33:51.498Z] 	at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:84)
[task 2020-10-22T21:33:51.498Z] 	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:59)
[task 2020-10-22T21:33:51.498Z] 	at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
[task 2020-10-22T21:33:51.498Z] 	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:38)
[task 2020-10-22T21:33:51.498Z] 	at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
[task 2020-10-22T21:33:51.498Z] 	at org.mozilla.fenix.share.ShareControllerTest.handleShareToApp should start a new sharing activity and close this(ShareControllerTest.kt:87)
[task 2020-10-22T21:33:51.498Z] 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
[task 2020-10-22T21:33:51.498Z] 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
[task 2020-10-22T21:33:51.498Z] 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
[task 2020-10-22T21:33:51.498Z] 	at java.lang.reflect.Method.invoke(Method.java:498)
[task 2020-10-22T21:33:51.498Z] 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
[task 2020-10-22T21:33:51.498Z] 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
[task 2020-10-22T21:33:51.498Z] 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
[task 2020-10-22T21:33:51.498Z] 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
[task 2020-10-22T21:33:51.498Z] 	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
[task 2020-10-22T21:33:51.498Z] 	at org.robolectric.RobolectricTestRunner$HelperTestRunner$1.evaluate(RobolectricTestRunner.java:546)
[task 2020-10-22T21:33:51.498Z] 	at org.robolectric.internal.SandboxTestRunner$2.lambda$evaluate$0(SandboxTestRunner.java:252)
[task 2020-10-22T21:33:51.498Z] 	at org.robolectric.internal.bytecode.Sandbox.lambda$runOnMainThread$0(Sandbox.java:89)
[task 2020-10-22T21:33:51.498Z] 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
[task 2020-10-22T21:33:51.498Z] 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[task 2020-10-22T21:33:51.498Z] 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
[task 2020-10-22T21:33:51.498Z] 	at java.lang.Thread.run(Thread.java:748)
```

--- a/app/src/main/java/org/mozilla/fenix/share/ShareController.kt
+++ b/app/src/main/java/org/mozilla/fenix/share/ShareController.kt
@@ -16,10 +16,11 @@ import android.net.Uri
 import androidx.annotation.VisibleForTesting
 import androidx.navigation.NavController
 import com.google.android.material.snackbar.Snackbar
+import kotlinx.coroutines.CoroutineDispatcher
 import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.GlobalScope
 import kotlinx.coroutines.Deferred
 import kotlinx.coroutines.Dispatchers
-import kotlinx.coroutines.GlobalScope
 import kotlinx.coroutines.launch
 import mozilla.components.concept.engine.prompt.ShareData
 import mozilla.components.concept.sync.Device
@@ -74,6 +75,7 @@ class DefaultShareController(
     private val navController: NavController,
     private val recentAppsStorage: RecentAppsStorage,
     private val viewLifecycleScope: CoroutineScope,
+    private val dispatcher: CoroutineDispatcher = Dispatchers.IO,
     private val dismiss: (ShareController.Result) -> Unit
 ) : ShareController {
 
@@ -88,7 +90,7 @@ class DefaultShareController(
     }
 
     override fun handleShareToApp(app: AppShareOption) {
-        viewLifecycleScope.launch(Dispatchers.IO) {
+        viewLifecycleScope.launch(dispatcher) {
             recentAppsStorage.updateRecentApp(app.activityName)
         }
 

--- a/app/src/test/java/org/mozilla/fenix/share/ShareControllerTest.kt
+++ b/app/src/test/java/org/mozilla/fenix/share/ShareControllerTest.kt
@@ -20,6 +20,7 @@ import io.mockk.verify
 import io.mockk.verifyOrder
 import kotlinx.coroutines.ExperimentalCoroutinesApi
 import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.test.TestCoroutineDispatcher
 import kotlinx.coroutines.test.TestCoroutineScope
 import mozilla.components.concept.engine.prompt.ShareData
 import mozilla.components.concept.sync.Device
@@ -28,11 +29,13 @@ import mozilla.components.concept.sync.TabData
 import mozilla.components.feature.accounts.push.SendTabUseCases
 import mozilla.components.feature.share.RecentAppsStorage
 import mozilla.components.support.test.robolectric.testContext
+import mozilla.components.support.test.rule.MainCoroutineRule
 import org.junit.After
 import org.junit.Assert.assertEquals
 import org.junit.Assert.assertNotEquals
 import org.junit.Assert.assertTrue
 import org.junit.Before
+import org.junit.Rule
 import org.junit.Test
 import org.junit.runner.RunWith
 import org.mozilla.fenix.R
@@ -62,6 +65,7 @@ class ShareControllerTest {
         TabData("title1", "url1")
     )
     private val textToShare = "${shareData[0].url}\n\n${shareData[1].url}"
+    private val testDispatcher = TestCoroutineDispatcher()
     private val testCoroutineScope = TestCoroutineScope()
     private val sendTabUseCases = mockk<SendTabUseCases>(relaxed = true)
     private val snackbar = mockk<FenixSnackbar>(relaxed = true)
@@ -70,8 +74,11 @@ class ShareControllerTest {
     private val recentAppStorage = mockk<RecentAppsStorage>(relaxed = true)
     private val controller = DefaultShareController(
         context, shareSubject, shareData, sendTabUseCases, snackbar, navController,
-        recentAppStorage, testCoroutineScope, dismiss
+        recentAppStorage, testCoroutineScope, testDispatcher, dismiss
     )
+
+    @get:Rule
+    val coroutinesTestRule = MainCoroutineRule(testDispatcher)
 
     @Before
     fun setUp() {
@@ -102,12 +109,13 @@ class ShareControllerTest {
         val activityContext: Context = mockk<Activity>()
         val testController = DefaultShareController(
             activityContext, shareSubject, shareData, mockk(),
-            mockk(), mockk(), recentAppStorage, testCoroutineScope, dismiss
+            mockk(), mockk(), recentAppStorage, testCoroutineScope, testDispatcher, dismiss
         )
         every { activityContext.startActivity(capture(shareIntent)) } just Runs
         every { recentAppStorage.updateRecentApp(appShareOption.activityName) } just Runs
 
         testController.handleShareToApp(appShareOption)
+        testDispatcher.advanceUntilIdle()
 
         // Check that the Intent used for querying apps has the expected structure
         assertTrue(shareIntent.isCaptured)
@@ -138,7 +146,7 @@ class ShareControllerTest {
         val activityContext: Context = mockk<Activity>()
         val testController = DefaultShareController(
             activityContext, shareSubject, shareData, mockk(),
-            snackbar, mockk(), recentAppStorage, testCoroutineScope, dismiss
+            snackbar, mockk(), recentAppStorage, testCoroutineScope, testDispatcher, dismiss
         )
         every { recentAppStorage.updateRecentApp(appShareOption.activityName) } just Runs
         every { activityContext.startActivity(capture(shareIntent)) } throws SecurityException()
@@ -166,7 +174,7 @@ class ShareControllerTest {
         val activityContext: Context = mockk<Activity>()
         val testController = DefaultShareController(
             activityContext, shareSubject, shareData, mockk(),
-            snackbar, mockk(), recentAppStorage, testCoroutineScope, dismiss
+            snackbar, mockk(), recentAppStorage, testCoroutineScope, testDispatcher, dismiss
         )
         every { recentAppStorage.updateRecentApp(appShareOption.activityName) } just Runs
         every { activityContext.startActivity(capture(shareIntent)) } throws ActivityNotFoundException()
@@ -316,6 +324,7 @@ class ShareControllerTest {
             mockk(),
             mockk(),
             mockk(),
+            mockk(),
             mockk()
         )
         val controllerWithMoreSharedTabs = controller
@@ -344,7 +353,7 @@ class ShareControllerTest {
         )
         val controller = DefaultShareController(
             context, shareSubject, shareData, sendTabUseCases, snackbar, navController,
-            recentAppStorage, testCoroutineScope, dismiss
+            recentAppStorage, testCoroutineScope, testDispatcher, dismiss
         )
 
         val expectedShareText = "${shareData[0].url}\n\nurl0\n\n${shareData[2].url}"
@@ -360,7 +369,7 @@ class ShareControllerTest {
     fun `getShareSubject will return a concatenation of tab titles if "shareSubject" is null`() {
         val controller = DefaultShareController(
             context, null, shareData, sendTabUseCases, snackbar, navController,
-            recentAppStorage, testCoroutineScope, dismiss
+            recentAppStorage, testCoroutineScope, testDispatcher, dismiss
         )
 
         assertEquals("title0, title1", controller.getShareSubject())

